lib/task.ml

```diff
@@ -1,33 +1,102 @@
+open EffectHandlers
+open EffectHandlers.Deep
+
 type 'a task = unit -> 'a
 
-type 'a promise = ('a, exn) result option Atomic.t
+type 'a promise_state =
+  Returned of 'a
+| Raised of exn
+| Pending of (unit, unit) continuation list
 
-exception TasksActive
+type 'a promise = 'a promise_state Atomic.t
 
-type task_msg =
-  Task : 'a task * 'a promise -> task_msg
-| Quit : task_msg
+type message =
+  Work of (unit -> unit)
+| Quit
 
 type pool_data = {
   domains : unit Domain.t array;
-  task_chan : task_msg Multi_channel.t;
+  task_chan : message Multi_channel.t;
   name: string option
 }
 
 type pool = pool_data option Atomic.t
 
-let do_task f p =
+let get_pool_data p =
+  match Atomic.get p with
+  | None -> raise (Invalid_argument "pool already torn down")
+  | Some p -> p
+
+let do_task pd f p =
   try
     let res = f () in
-    Atomic.set p (Some (Ok res))
+    match Atomic.exchange p (Returned res) with
+    | Pending l ->
+        List.iter (fun k -> Multi_channel.send pd.task_chan (Work (continue k))) l
+    | _ -> failwith "Task.do_task: impossible (1)"
   with e ->
-    Atomic.set p (Some (Error e));
-    match e with
-    | TasksActive -> raise e
-    | _ -> ()
+    begin match Atomic.exchange p (Raised e) with
+    | Pending l -> List.iter (fun k ->
+        Multi_channel.send pd.task_chan (Work (fun _ -> discontinue k e))) l
+    | _ -> failwith "Task.do_task: impossible (2)"
+    end
 
-let named_pools = Hashtbl.create 8
+let async pool f =
+  let pd = get_pool_data pool in
+  let p = Atomic.make (Pending []) in
+  Multi_channel.send pd.task_chan (Work (fun _ -> do_task pd f p));
+  p
+
+type _ eff += Wait : 'a promise -> unit eff
 
+let rec await promise =
+  match Atomic.get promise with
+  | Returned v -> v
+  | Raised e -> raise e
+  | Pending _ -> perform (Wait promise); await promise
+
+let step (type a) (f : a -> unit) (v : a) : unit =
+  try_with f v
+  { effc = fun (type a) (e : a eff) ->
+      match e with
+      | Wait p -> Some (fun (k : (a, _) continuation) ->
+          let rec loop () =
+            let old = Atomic.get p in
+            match old with
+            | Pending l ->
+                if Atomic.compare_and_set p old (Pending (k::l)) then ()
```

Comment:
Will this result in LIFO execution of awaiters? Does it matter?

Reply:
The ordering doesn't matter here. It is intentionally left unspecified to give the implementation freedom to change it later.